SofaKernel/modules/SofaBaseMechanics/DiagonalMass.inl

```diff
@@ -528,55 +532,283 @@ void DiagonalMass<DataTypes, MassType>::getElementMass(unsigned int index, defau
+        if (this->mstate && d_vertexMass.getValue().size() > 0 && d_vertexMass.getValue().size() < (unsigned)this->mstate->getSize())
+        {
+            MassVector &masses= *d_vertexMass.beginEdit();
+            size_t i = masses.size()-1;
+            size_t n = (size_t)this->mstate->getSize();
+            masses.reserve(n);
+            while (masses.size() < n)
+                masses.push_back(masses[i]);
+            d_vertexMass.endEdit();
+        }
+
+        massInitialization();
+    }
+}
+
+
+template <class DataTypes, class MassType>
+void DiagonalMass<DataTypes, MassType>::massInitialization()
+{
+    //Mass initialization process
+    if(d_vertexMass.isSet() || d_massDensity.isSet() || d_totalMass.isSet() )
+    {
+        //totalMass data is prioritary on vertexMass and massDensity
+        if (d_totalMass.isSet())
+        {
+            if(d_vertexMass.isSet() || d_massDensity.isSet())
+            {
+                msg_warning(this) << "totalMass value overriding other mass information (vertexMass or massDensity).\n"
+                                  << "To remove this warning you need to define only one single mass information data field.";
+            }
+            checkTotalMassInit();
+            initFromTotalMass();
+        }
+        //massDensity is secondly considered
+        else if(d_massDensity.isSet())
+        {
+            if(d_vertexMass.isSet())
+            {
+                msg_warning(this) << "massDensity value overriding the value of the attribute vertexMass.\n"
+                                  << "To remove this warning you need to set either vertexMass or massDensity data field, but not both.";
+            }
+            if(!checkMassDensity())
+            {
+                checkTotalMassInit();
+                initFromTotalMass();
+            }
+            else
+            {
+                initFromMassDensity();
+            }
+        }
+        //finally, the vertexMass is used
+        else if(d_vertexMass.isSet())
+        {
+            if(!checkVertexMass())
+            {
+                checkTotalMassInit();
+                initFromTotalMass();
+            }
+            else
+            {
+                initFromVertexMass();
+            }
+        }
+    }
+    // if no mass information provided, default initialization uses totalMass
+    else
+    {
+        msg_info() << "No information about the mass is given." << msgendl
+                      "Default : totalMass = 1.0";
+        checkTotalMassInit();
+        initFromTotalMass();
+    }
+
+    //Info post-init
+    msg_info() << "mass information computed :" << msgendl
+               << "totalMass   = " << d_totalMass.getValue() << msgendl
+               << "massDensity = " << d_massDensity.getValue() << msgendl
+               << "vertexMass  = " << d_vertexMass.getValue();
```

Comment:
Will this print the complete vector of vertex masses? Maybe instead we could print the min,max,avg values?